pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,995 @@
+
+			let index = schedule_ids
+				.iter()
+				.position(|x| *x == schedule_id)
+				.ok_or(Error::<T>::NoPlannedExecutionFoundOnBlock)?;
+
+			schedule_ids.remove(index);
+
+			if schedule_ids.is_empty() {
+				*maybe_schedule_ids = None;
+			}
+			Ok(())
+		})?;
+
+		Ok(())
+	}
+
+	fn suspend_schedule(owner: &T::AccountId, schedule_id: ScheduleId) -> DispatchResult {
+		Suspended::<T>::insert(schedule_id, ());
+		Self::deposit_event(Event::Suspended {
+			id: schedule_id,
+			who: owner.clone(),
+		});
+
+		Ok(())
+	}
+
+	fn convert_to_currency_if_asset_is_not_native(
+		asset_id: T::Asset,
+		asset_amount: u128,
+	) -> Result<u128, DispatchError> {
+		let amount = if asset_id == T::NativeAssetId::get() {
+			asset_amount
+		} else {
+			let price = Self::get_price_from_last_block_oracle(T::NativeAssetId::get(), asset_id)?;
+
+			price.checked_mul_int(asset_amount).ok_or(ArithmeticError::Overflow)?
+		};
+
+		Ok(amount)
+	}
+
+	fn remove_schedule_from_storages(owner: &T::AccountId, schedule_id: ScheduleId) {
+		Schedules::<T>::remove(schedule_id);
+		Suspended::<T>::remove(schedule_id);
+		ScheduleOwnership::<T>::remove(owner, schedule_id);
+		RemainingRecurrences::<T>::remove(schedule_id);
+		RemainingAmounts::<T>::remove(schedule_id);
+	}
+
+	fn remove_planning_or_suspension(
+		schedule_id: ScheduleId,
+		next_execution_block: Option<T::BlockNumber>,
+	) -> DispatchResult {
+		match next_execution_block {
+			Some(block) => {
+				let schedule_ids_on_block = ScheduleIdsPerBlock::<T>::get(block);
+
+				ensure!(
+					schedule_ids_on_block.contains(&schedule_id),
+					Error::<T>::NoPlannedExecutionFoundOnBlock,
+				);
+
+				Self::remove_schedule_id_from_next_execution_block(schedule_id, block)?;
+			}
+			None => {
+				Self::ensure_schedule_is_suspended(schedule_id)?;
+				Suspended::<T>::remove(schedule_id);
+			}
+		};
+
+		Ok(())
+	}
+
+	fn complete_dca(owner: &T::AccountId, schedule_id: ScheduleId) {
+		Self::remove_schedule_from_storages(owner, schedule_id);
+		Self::deposit_event(Event::Completed {
+			id: schedule_id,
+			who: owner.clone(),
+		});
+	}
+}
+
+pub trait RandomnessProvider {
+	fn generator() -> StdRng;
+}
+
+impl<T: Config> RandomnessProvider for Pallet<T> {
+	fn generator() -> StdRng {
+		let hash_value = pallet_relaychain_info::Pallet::<T>::parent_hash();
+		let mut seed_arr = [0u8; 8];
+		seed_arr.copy_from_slice(&hash_value.as_fixed_bytes()[0..8]);
+		let seed = u64::from_le_bytes(seed_arr);
+		rand::rngs::StdRng::seed_from_u64(seed)
+	}
+}
+
+#[macro_export]
+macro_rules! exec_or_return_if_none {
+	($opt:expr) => {
+		match $opt {
+			Some(val) => val,
+			None => {
+				log::error!(target: "runtime::dca", "Unexpected error happened while executing schedule.");
+				return;
+			}
+		}
+	};
+}
+
+#[macro_export]
+macro_rules! exec_or_return_if_err {
+	($res:expr) => {
+		match $res {
+			Ok(val) => val,
+			Err(e) => {
+				log::error!(
+					target: "runtime::dca",
```

Comment:
logging is not sufficient resolution to error, the whole schedule execution has to be reverted, we cannot end up with some inconsistent state